mobile/src/main/java/ch/epfl/sdp/mobile/ui/play/ExpandableFloatingActionButton.kt

```diff
@@ -0,0 +1,240 @@
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.util.lerp
+import ch.epfl.sdp.mobile.ui.FadeEnterTransition
+import ch.epfl.sdp.mobile.ui.FadeExitTransition
+import ch.epfl.sdp.mobile.ui.PawniesColors.Beige050
+import ch.epfl.sdp.mobile.ui.PawniesColors.Green100
+import ch.epfl.sdp.mobile.ui.PawniesColors.Green800
+
+/** The state of an [ExpandableFloatingActionButton] */
+interface ExpandableFloatingActionButtonState {
+
+  /** True iff the button should be currently expanded. */
+  var expanded: Boolean
+}
+
+/**
+ * An interface representing the colors of an [ExpandableFloatingActionButton]. The colors may
+ * change depending on whether the button is currently expanded or not.
+ */
+interface ExpandableFloatingActionButtonColors {
+
+  /**
+   * Returns a state with the color for the button background.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun background(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button border.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun border(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button content color.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun contentColor(expanded: Boolean): State<Color>
+}
+
+/**
+ * A composable which displays a floating action which may be expanded.
+ *
+ * @param expandedContent the content to display when expanded. Usually a list of
+ * [ExpandableFloatingActionButtonItem].
+ * @param modifier the [Modifier] for this composable.
+ * @param state the backing state for this button.
+ * @param colors the colors for this button.
+ * @param elevation the elevation for this button.
+ * @param interactionSource the [MutableInteractionSource] to keep track of inputs.
+ * @param expandedCornerSize the size of the rounded corners when the button is expanded.
+ * @param collapsedContent the content to display when collapsed.
+ */
+@OptIn(
+    ExperimentalAnimationApi::class,
+    ExperimentalMaterialApi::class,
+)
+@Composable
+fun ExpandableFloatingActionButton(
+    expandedContent: @Composable ColumnScope.() -> Unit,
+    modifier: Modifier = Modifier,
+    state: ExpandableFloatingActionButtonState = rememberExpandableFloatingActionButtonState(),
+    colors: ExpandableFloatingActionButtonColors = DefaultExpandableFloatingActionButtonColors,
+    elevation: ButtonElevation = ButtonDefaults.elevation(),
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
+    expandedCornerSize: Dp = 8.dp,
+    collapsedContent: @Composable RowScope.() -> Unit,
+) {
+  val color by colors.background(state.expanded)
+  val border by colors.border(state.expanded)
+  val contentColor by colors.contentColor(state.expanded)
+  val progress = animateFloatAsState(if (state.expanded) 1f else 0f)
```

Comment:
[Question] Does this mean that the progress will take values from 1f to 0f when `expanded ` is `false` ? If so, why is it a `val` and not a `var` ?